openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,131 @@
+as discussable, in which case a discussion UI will show up in-context.
+
+Currently this will only be supported for the edX Forums internal
+discussion provider, but it could be extended in the future to support
+other providers.
+
+In addition to providing a way to mark Units as discussable, there will
+also be a way to simply enable discussion for all units in the course.
+One further control provided here will be to toggle discussions on or
+off for graded discussions.
+
+Finally, it will also be possible to group discussions at a Sequence
+level instead of the Unit level.
+
+
+Requirements
+------------
+
+In order to enable in-context discussions with the above features we
+need a way to:
+
+- Mark a Unit as discussable
+- Store configuration options for discussions
+- Automatically mark/unmark units as discussable when settings change
+- Associate a unit with a corresponding discussion id
+- Make this discussion ID available via an API
+
+
+Consideration
+-------------
+
+Most of these configuration entries would be right at home in the
+`DiscussionsConfiguration` model in `plugin_settings`, however since they need
+to be available during course import-export, they should be stored in the
+course object itself.
+
+There is already a way to associate a discussion ID with a Discussion XBlock
+using its usage key. This same mechanism can be used to associate a Unit usage
+key with a corresponding discussion id.
+
+However the current mechanism has a few issues. It is stored as a JSON
+structure in the `DiscussionsIdMapping` model which has course id and a mapping
+of the discussion id to the xblock usage key in a single dict.
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course structure with the following structure:
+
+.. code-block:: JSON
+
+    {
+        "edx-next": {
+            "discussions_enable_in_context": bool,
+            "discussions_enable_graded_units": bool,
+            "discussions_group_at_subsection": bool,
+            "discussions_custom_visibility": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
```

Comment:
The multiple providers question is also an interesting one as it relates to import/export, because when you export your course from one instance and import it into another (like MIT might), those instances might have different available providers. What's the right thing to try in that scenario? Should a configuration only be valid for one provider, and the course team make the explicit decision to enable another provider? Or should it just work, even if it means a different provider is being set up by default? I'd favor the latter, but that's more of a product question.

It's not that we have to code that failover behavior, but the desired behavior has implications for how the configuration is organized.

Reply:
This configuraiton is only part of the puzzle. The main configuration link is stored in the DB so won't be imported-exported and the provider will need to be set up manually for each import.

Reply:
And each re-run as well? Won't that annoy course teams?

Reply:
I'd like to clarify on this point. The link between a discussion  provider and the course key is stored in the db, so for each new course run, you'd need to go through the provider setup again, and enter in the LTI credentials for each run, or import. 
However, settings for the internal forums are still part of exports, so if you are using that you just need to select the provider, click next, and you're done. 

So in my approach above for instance, you could set up an internal forum config, so if a provider you want isn't available for some reason (these are all LTI providers so it wouldn't be due to a  missing plugin) you an always switch back to the internal forums. 

The settings here are currently only for the internal forums, and LTI only has settings for LTI credential and launch URL, so if you need to change the provider, there isn't much you can carry forward anyway. 

That said, I think it's a good idea to move common setting to a common location, which will ease setting things up when changing providers.